### ISSUE TITLE
🛠️ Enforce descriptions on `intl` strings

### ISSUE THREAD

**Issue by petertgiles** [GCTC-NTGC/gc-digital-talent#3667]
While reviewing a translation PR, Matt observed that it is difficult to review a string with no description or even find where it came from.  Descriptions can help us get better quality translations back, too.

There is an eslint rule to enforce having descriptions on strings.  Should we enable it?
https://formatjs.io/docs/tooling/linter/#enforce-description

**Comment by petertgiles** [comment]
Now that we have IDs on strings in the code the issue with finding the string origin is moot.  The points about reviewing translations and getting better quality translations with a description are still valid though, I think.

**Comment by patcon** [comment]
No block, but soft -1.

I feel like we already over-use descriptions on principle.

1. make its harder to de-dup single word translations into single string
2. description are often just a verbose representation of hierarchy, which feels like it would be clearer as a custom id. "The header text of the foo section on the bar page" => `bar.foo.header`. 

tl;dr - I already feel we use descriptions too often, with negligible value

**Comment by petertgiles** [comment]
Hehe, I think we need a "controversial" tag!

**Comment by mnigh** [comment]
> 2\. description are often just a verbose representation of hierarchy, which feels like it would be clearer as a custom id. "The header text of the foo section on the bar page" => `bar.foo.header`.

As developers, we might not see the value of the descriptions, but more context for a translator can be helpful when translating something, especially single words that are spelled the same but have different meanings (homonyms).

**Comment by patcon** [comment]
@mnigh you're right, i don't want to be too dev-centered, and dot-notation isn't friendly. But I'd still say we're overstating the value of many of these descriptions as-is, and their cost in reducing our ability to de-dup strings.

Counter-proposal might be to standardize on descriptions that closely match a structure of their manual IDs, like:
- `[page] > [section] > [obection]`
- `Bar page > Foo section > header label` (for e.g., `bar.foo.header`)
- `All pages > submit button label` (for e.g., `.all.submit`)

These would promote the same hierarchical thought from devs (and maybe designers) that would hopefully help us dedup more strings (when we recognize the re-use patterns of the strings) :)

Again, no strong feelings here, so if this doesn't resonate with anyone else, I retract :)

**Comment by petertgiles** [comment]
I agree that we're probably duplicating way too many strings.  For example, we have "date de fin" five times in the talent search fr file.  :laughing:   Of course a phrase might be translated differently in a different context but I'm guessing that should be the exception rather than the rule.

Now that the process is that devs must have their translations ready for every PR, I think it incentivizes us to reuse strings rather than blindly make new ones and have to request translation later.  I know I've started reusing them when possible rather than have to send away for new strings.

That said, I think the with increased reuse the description of the string becomes even more important to know when it can be reused.  I hope we can start using more general descriptions to promote reuse.

**Comment by mnigh** [comment]
@patcon the value of the descriptions depends on how they are written, and we don't have a standard way one doing so. the [formatjs docs](https://formatjs.io/docs/react-intl/api/#message-descriptor) describe the purpose of the description:
> **description**: Context for the translator about how it's used in the UI

i agree with standardizing the way descriptions are written. i wonder if this would be something the designers could provide since they are providing the copy itself and should be able to describe "how it's used in the UI"?

@petertgiles reusing strings is great, but can get tricky when trying to take parts of strings out of context. i'm not totally against it (as we already do this, see next sentence), but there are considerations like punctuation (french colon spacing) that can make reuse tricky. having said this, we already have a pattern to reuse strings, which is adding them to one of the `messages` files in [/frontend/common/src/messages](https://github.com/GCTC-NTGC/gc-digital-talent/tree/main/frontend/common/src/messages).

**Comment by patcon** [comment]
> Now that the process is that devs must have their translations ready for every PR, I think it incentivizes us to reuse strings rather than blindly make new ones and have to request translation later. -- @petertgiles 

Good call, that's a new incentive, and might change things even if nothing else changes :)

> i wonder if this would be something the designers could provide -- @mnigh 

down for this! but also don't want to bottleneck more on design team, esp if having us think it thru would help us all connect dots throughout our pipeline in a good way

still +1 to more re-use. my general loose vote is for trying and seeing if any downsides materialize, rather than avoiding based on worry they might

e.g., why not have a "colon" string and always add that after? it can be its own thing, and doesn't need to fragment other string purposes. We already have the option to do this for "ellipses", though not sure we take advantage of it fully

https://github.com/GCTC-NTGC/gc-digital-talent/blob/99406a9f73994cbf43c8eaabb2bb4efadf9cffe8/frontend/talentsearch/src/js/components/workPreferencesForm/WorkPreferencesForm.tsx#L54-L57

https://github.com/GCTC-NTGC/gc-digital-talent/blob/99406a9f73994cbf43c8eaabb2bb4efadf9cffe8/frontend/talentsearch/src/js/lang/fr.json#L2-L4

**Comment by vd1992** [comment]
definitely got a bunch of strings around where the descriptions is just filler

**Comment by patcon** [comment]
vachan, tristan, patcon not feeling support for this as high-value. very repetitive. not writing good enough descriptions to be helpful, just description of location in doc.

matt feels that enforcing descriptions is separate from description quality. something better than nothing.

**Comment by patcon** [comment]
matt: halfway solutions seem not good. don't see benefit to doing it half the time. vachan: disagree.

patcon: i feel ok with halfway.

tristan: willing to defer judgement to matt.

#todo better internal guidelines for writing descriptions.

**Comment by mnigh** [comment]
- the descriptions are not for us as developers, they are meant to provide translators "context for the translator about how it's used in the UI"
- i agree with peter, "good description > bad description > no description"
- an issue has been created to create internal guidelines for writing descriptions, but i think this doesn't replace enforcing descriptions: 
https://github.com/GCTC-NTGC/gc-digital-talent/issues/5106
